src/crystal/system/win32/iocp.cr

```diff
@@ -168,8 +168,16 @@ module Crystal::IOCP
     end
   end
 
-  def self.overlapped_operation(target, handle, method, timeout, *, writing = false, &)
+  def self.overlapped_operation(file_descriptor, method, timeout, *, writing = false, &)
+    handle = file_descriptor.windows_handle
+    seekable = LibC.SetFilePointerEx(handle, 0, out original_offset, IO::Seek::Current) != 0
+
     OverlappedOperation.run(handle) do |operation|
+      overlapped = operation.to_unsafe
+      if seekable
+        overlapped.value.union.offset.offset = LibC::DWORD.new(original_offset)
+        overlapped.value.union.offset.offsetHigh = LibC::DWORD.new(original_offset >> 32)
```

Comment:
**issue:** `LibC::DWORD.new(original_offset)` would raise if `original_offset` is too big to fit into 32 bits. That would render the `offsetHigh` part useless.
We should use the wrapping constructor `LibC::DWORD.new!` instead. Also for `offsetHigh` because we can be sure that `original_offset >> 32` fits into `DWORD`.
```suggestion
        overlapped.value.union.offset.offset = LibC::DWORD.new!(original_offset)
        overlapped.value.union.offset.offsetHigh = LibC::DWORD.new!(original_offset >> 32)
```